Unstable test TestTruncatePartitionWithGlobalIndex, related to InfoSchema v2

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```sh
(cd pkg/ddl/tests/partition/ ; go test -v -run TestTruncatePartitionWithGlobalIndex -tags=intest -count 1000 -failfast)
```

### 2. What did you expect to see? (Required)
Success

### 3. What did you see instead (Required)
Failure, part of log:
```
[2024/05/31 13:03:04.125 +02:00] [INFO] [session.go:3959] ["CRUCIAL OPERATION"] [conn=96] [schemaVersion=64] [cur_db=test] [sql="alter table test_global truncate partition p2;"] [user=]
[2024/05/31 13:03:04.125 +02:00] [INFO] [ddl_worker.go:261] ["add DDL jobs"] [category=ddl] ["batch count"=1] [jobs="ID:111, Type:truncate partition, State:queueing, SchemaState:public, SchemaID:2, TableID:104, RowCount:0, ArgLen:2, start time: 2024-05-31 13:03:04.125 +0200 CEST, Err:<nil>, ErrCount:0, SnapshotVersion:0, LocalMode: false; "] [table=true]
[2024/05/31 13:03:04.125 +02:00] [INFO] [ddl.go:1192] ["start DDL job"] [category=ddl] [job="ID:111, Type:truncate partition, State:queueing, SchemaState:public, SchemaID:2, TableID:104, RowCount:0, ArgLen:2, start time: 2024-05-31 13:03:04.125 +0200 CEST, Err:<nil>, ErrCount:0, SnapshotVersion:0, LocalMode: false"] [query="alter table test_global truncate partition p2;"]
[2024/05/31 13:03:04.126 +02:00] [INFO] [set.go:170] ["set global var"] [conn=98] [name=tidb_schema_cache_size] [val=1073741824]
[2024/05/31 13:03:04.127 +02:00] [INFO] [ddl_worker.go:1210] ["run DDL job"] [worker="worker 196, tp general"] [category=ddl] [jobID=111] [conn=96] [category=ddl] [job="ID:111, Type:truncate partition, State:queueing, SchemaState:public, SchemaID:2, TableID:104, RowCount:0, ArgLen:0, start time: 2024-05-31 13:03:04.125 +0200 CEST, Err:<nil>, ErrCount:0, SnapshotVersion:0, LocalMode: false"]
[2024/05/31 13:03:04.128 +02:00] [ERROR] [domain.go:300] ["failed to load schema diff"] [error="builder's (v2=true) infoschema mismatch, return error to trigger full reload"] [errorVerbose="builder's (v2=true) infoschema mismatch, return error to trigger full reload\ngithub.com/pingcap/tidb/pkg/infoschema.(*Builder).InitWithOldInfoSchema\n\t/home/user/tidb/pkg/infoschema/builder.go:799\ngithub.com/pingcap/tidb/pkg/domain.(*Domain).tryLoadSchemaDiffs\n\t/home/user/tidb/pkg/domain/domain.go:477\ngithub.com/pingcap/tidb/pkg/domain.(*Domain).loadInfoSchema\n\t/home/user/tidb/pkg/domain/domain.go:283\ngithub.com/pingcap/tidb/pkg/domain.(*Domain).Reload\n\t/home/user/tidb/pkg/domain/domain.go:624\ngithub.com/pingcap/tidb/pkg/domain.(*Domain).loadSchemaInLoop\n\t/home/user/tidb/pkg/domain/domain.go:937\ngithub.com/pingcap/tidb/pkg/domain.(*Domain).Init.func3\n\t/home/user/tidb/pkg/domain/domain.go:1305\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupEnhancedWrapper).Run.func1\n\t/home/user/tidb/pkg/util/wait_group_wrapper.go:99\nruntime.goexit\n\t/home/user/go/src/runtime/asm_amd64.s:1650"] [stack="github.com/pingcap/tidb/pkg/domain.(*Domain).loadInfoSchema\n\t/home/user/tidb/pkg/domain/domain.go:300\ngithub.com/pingcap/tidb/pkg/domain.(*Domain).Reload\n\t/home/user/tidb/pkg/domain/domain.go:624\ngithub.com/pingcap/tidb/pkg/domain.(*Domain).loadSchemaInLoop\n\t/home/user/tidb/pkg/domain/domain.go:937\ngithub.com/pingcap/tidb/pkg/domain.(*Domain).Init.func3\n\t/home/user/tidb/pkg/domain/domain.go:1305\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupEnhancedWrapper).Run.func1\n\t/home/user/tidb/pkg/util/wait_group_wrapper.go:99"]
[2024/05/31 13:03:04.138 +02:00] [INFO] [domain.go:324] ["full load InfoSchema success"] [currentSchemaVersion=64] [neededSchemaVersion=65] ["start time"=10.388472ms]
    result.go:160:
                Error Trace:    /home/user/tidb/pkg/testkit/result.go:160
                                                        /home/user/tidb/pkg/ddl/tests/partition/db_partition_test.go:1459
                Error:          Not equal:
                                expected: true
                                actual  : false
                Test:           TestTruncatePartitionWithGlobalIndex
                Messages:       the result doesn't contain the exepected IndexRangeScan
                                Point_Get 1.00 root table:test_global, index:idx_b(b)
[2024/05/31 13:03:04.139 +02:00] [INFO] [ddl.go:1230] ["DoDDLJob will quit because context done"] [category=ddl]
[2024/05/31 13:03:04.139 +02:00] [INFO] [mock.go:100] ["owner manager retire owner"] [category=ddl] [ID=09af9669-4962-47e5-a658-a3f6ed6a894c] [ownerKey=/tidb/ddl/fg/owner]
[2024/05/31 13:03:04.139 +02:00] [INFO] [schema_version.go:376] ["wait latest schema version encounter error"] [category=ddl] [ver=65] [jobID=111] ["take time"=11.357062ms] [error="context canceled"] [errorVerbose="context canceled\ngithub.com/pingcap/errors.AddStack\n\t/home/user/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/errors.go:178\ngithub.com/pingcap/errors.Trace\n\t/home/user/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/juju_adaptor.go:15\ngithub.com/pingcap/tidb/pkg/ddl.(*MockSchemaSyncer).OwnerCheckAllVersions\n\t/home/user/tidb/pkg/ddl/mock.go:132\ngithub.com/pingcap/tidb/pkg/ddl.checkAllVersions\n\t/home/user/tidb/pkg/ddl/schema_version.go:374\ngithub.com/pingcap/tidb/pkg/ddl.waitSchemaChanged\n\t/home/user/tidb/pkg/ddl/ddl_worker.go:1438\ngithub.com/pingcap/tidb/pkg/ddl.(*jobScheduler).delivery2Worker.func1\n\t/home/user/tidb/pkg/ddl/job_table.go:577\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1\n\t/home/user/tidb/pkg/util/wait_group_wrapper.go:157\nruntime.goexit\n\t/home/user/go/src/runtime/asm_amd64.s:1650"]
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
current master:
```
tidb_version(): Release Version: v8.2.0-alpha-277-g90e1049b67
Edition: Community
Git Commit Hash: 90e1049b6768120b66b02cae68f0fab0d36320cd
Git Branch: master
UTC Build Time: 2024-05-31 11:07:48
GoVersion: go1.21.1
Race Enabled: false
Check Table Before Drop: false
Store: unistore
```

--- a/pkg/ddl/tests/partition/db_partition_test.go
+++ b/pkg/ddl/tests/partition/db_partition_test.go
@@ -1412,7 +1412,7 @@ func TestGlobalIndexUpdateInDropPartition(t *testing.T) {
 }
 
 func TestTruncatePartitionWithGlobalIndex(t *testing.T) {
-	store := testkit.CreateMockStore(t)
+	store, dom := testkit.CreateMockStoreAndDomain(t)
 	tk := testkit.NewTestKit(t, store)
 	tk.MustExec("use test")
 	tk.MustExec("set tidb_enable_global_index=true")
@@ -1436,6 +1436,8 @@ func TestTruncatePartitionWithGlobalIndex(t *testing.T) {
 	tk2.MustExec(`use test`)
 	tk2.MustExec(`begin`)
 	tk2.MustExec(`insert into test_global values (5,5,5)`)
+
+	v1 := dom.InfoSchema().SchemaMetaVersion()
 	syncChan := make(chan bool)
 	go func() {
 		tk.MustExec("alter table test_global truncate partition p2;")
@@ -1447,9 +1449,13 @@ func TestTruncatePartitionWithGlobalIndex(t *testing.T) {
 			tk4.MustExec(`use test`)
 			res := tk4.MustQuery(`admin show ddl jobs where db_name = 'test' and table_name = 'test_global' and job_type = 'truncate partition'`).Rows()
 			if len(res) == 1 && res[0][i] == s {
-				break
+				v2 := dom.InfoSchema().SchemaMetaVersion()
+				if v2 > v1 {
+					// Also wait for the new infoschema loading
+					break
+				}
 			}
-			time.Sleep(100 * time.Millisecond)
+			time.Sleep(10 * time.Millisecond)
 		}
 	}
 	waitFor(4, "delete only")

--- a/pkg/ddl/tests/partition/db_partition_test.go
+++ b/pkg/ddl/tests/partition/db_partition_test.go
@@ -1449,7 +1449,7 @@ func TestTruncatePartitionWithGlobalIndex(t *testing.T) {
 			if len(res) == 1 && res[0][i] == s {
 				break
 			}
-			time.Sleep(10 * time.Millisecond)
+			time.Sleep(100 * time.Millisecond)
 		}
 	}
 	waitFor(4, "delete only")

